src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1052,15 +1048,16 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
                 var newIdToProjectStateMap = this.SolutionState.ProjectStates;
                 var newIdToTrackerMap = _projectIdToTrackerMap;
 
-                foreach (var (doc, tree) in builder)
+                foreach (var siblingId in allDocumentIds)
                 {
                     // if we don't have one or it is stale, create a new partial solution
-                    var tracker = this.GetCompilationTracker(doc.Id.ProjectId);
-                    var newTracker = tracker.FreezePartialStateWithTree(this, doc, tree, cancellationToken);
+                    var tracker = this.GetCompilationTracker(siblingId.ProjectId);
+                    var siblingDocumentState = this.SolutionState.GetRequiredDocumentState(siblingId);
+                    var newTracker = tracker.FreezePartialStateWithTree(this, siblingDocumentState, syntaxTree, cancellationToken);
```

Comment:
>syntaxTree [](http://example.com/codeflow?start=100&length=10)

Can you add a comment on why this is safe to do? I thought you had said this was only safe if there weren't any directive differences.